lib/Runtime/Library/JavascriptMap.cpp

```diff
@@ -6,462 +6,678 @@
+    {
+        return false;
+    }
 
-        int size = map->Size();
+    return this->DeleteFromVarMap<false>(simpleVar);
+}
+
+bool
+JavascriptMap::Delete(Var key)
+{
+    JS_REENTRANCY_LOCK(jsReentLock, this->GetScriptContext()->GetThreadContext());
+    switch (this->kind)
+    {
+    case MapKind::EmptyMap:
+        return false;
 
-        return JavascriptNumber::ToVar(size, scriptContext);
+    case MapKind::SimpleVarMap:
+        return this->DeleteFromSimpleVarMap(key);
+    case MapKind::ComplexVarMap:
+        return this->DeleteFromVarMap<true>(key);
+    default:
+        Assume(UNREACHED);
+        return false;
     }
+}
 
-    Var JavascriptMap::EntryEntries(RecyclableObject* function, CallInfo callInfo, ...)
+bool
+JavascriptMap::Get(Var key, Var* value)
+{
+    JS_REENTRANCY_LOCK(jsReentLock, this->GetScriptContext()->GetThreadContext());
+    switch (this->kind)
     {
-        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
-
-        ARGUMENTS(args, callInfo);
-        ScriptContext* scriptContext = function->GetScriptContext();
+    case MapKind::EmptyMap:
+        return false;
 
-        if (!JavascriptMap::Is(args[0]))
+    case MapKind::SimpleVarMap:
+    {
+        // First check if the key is in the map
+        MapDataNode* node = nullptr;
+        if (this->u.simpleVarMap->TryGetValue(key, &node))
         {
-            JavascriptError::ThrowTypeErrorVar(scriptContext, JSERR_NeedObjectOfType, _u("Map.prototype.entries"), _u("Map"));
+            *value = node->data.Value();
+            return true;
+        }
+        // If the key isn't in the map, check if the canonical value is
+        Var simpleVar = JavascriptConversion::TryCanonicalizeAsSimpleVar(key);
+        // If the simple var is the same as the original key, we know it isn't in the map
+        if (!simpleVar || simpleVar == key)
+        {
+            return false;
         }
 
-        JavascriptMap* map = JavascriptMap::FromVar(args[0]);
-
-        return scriptContext->GetLibrary()->CreateMapIterator(map, JavascriptMapIteratorKind::KeyAndValue);
+        if (!this->u.simpleVarMap->TryGetValue(simpleVar, &node))
+        {
+            return false;
+        }
+        *value = node->data.Value();
+        return true;
     }
-
-    Var JavascriptMap::EntryKeys(RecyclableObject* function, CallInfo callInfo, ...)
+    case MapKind::ComplexVarMap:
     {
-        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+        MapDataNode * node = nullptr;
+        if (!this->u.complexVarMap->TryGetValue(key, &node))
+        {
+            return false;
+        }
+        *value = node->data.Value();
+        return true;
+    }
+    default:
+        Assume(UNREACHED);
+        return false;
+    }
+}
 
-        ARGUMENTS(args, callInfo);
-        ScriptContext* scriptContext = function->GetScriptContext();
 
-        if (!JavascriptMap::Is(args[0]))
+bool
+JavascriptMap::Has(Var key)
```

Comment:
For discussion: ContainsKey and TryGetValue only differ in terms of return. 
Since the logic of Get and Has is same should we just use Get and return the result ignoring the value?

Reply:
i tend to agree, but it's only a few lines of code and there might be some extra cost from those stores and bigger code might make inlining less aggressive... obviously small costs, but the code duplication win is also pretty small so i'm not sure it's worth it
if it gets any more complicated we should definitely merge these paths though

---
In reply to: [174324761](https://github.com/Microsoft/ChakraCore/pull/4816#discussion_r174324761) [](ancestors = 174324761)